### PR TITLE
Pma/ns frac

### PR DESCRIPTION
**Pull request by pmattila**

### CODE REVIEW

**Review by mcb30 — CHANGES_REQUESTED** [review]
Based on previous attempts to get linuxptp code merged upstream, I think we need to ensure that each patch is minimal.  Further comments inline.

File: sk.c [hunk 1]
```diff
@@ -304,8 +304,11 @@ int sk_receive(int fd, void *buf, int buflen,
 	struct cmsghdr *cm;
 	struct iovec iov = { buf, buflen };
 	struct msghdr msg;
-	struct timespec *sw, *ts = NULL;
+	struct timespec *sw;
+	struct timespec *ts = NULL;
```
Comment (mcb30) [inline sk.c:309]: Probably best to leave these two lines as-is, or split out to a separate patch with a justification for why it should be changed.

File: sk.c [hunk 2]
```diff
@@ -341,10 +344,15 @@ int sk_receive(int fd, void *buf, int buflen,
 		level = cm->cmsg_level;
 		type  = cm->cmsg_type;
 		if (SOL_SOCKET == level && SO_TIMESTAMPING == type) {
-			if (cm->cmsg_len < sizeof(*ts) * 3) {
+			if (cm->cmsg_len < 3 * sizeof(struct timespec)) {
```
Comment (mcb30) [inline sk.c:348]: This should probably be left unaltered

File: sk.c [hunk 3]
```diff
@@ -341,10 +344,15 @@ int sk_receive(int fd, void *buf, int buflen,
...
 				pr_warning("short SO_TIMESTAMPING message");
 				return -1;
 			}
+			if (cm->cmsg_len >= 3 * sizeof(struct timespec) +
```
Comment (mcb30) [inline sk.c:352]: "sizeof(*ts) * 3 + sizeof(*hr)" to match existing code style

File: sk.c [hunk 4]
```diff
@@ -360,22 +368,22 @@ int sk_receive(int fd, void *buf, int buflen,
 	if (addr)
 		addr->len = msg.msg_namelen;
 
-	if (!ts) {
```
Comment (mcb30) [inline sk.c:351]: Should probably be left in place, or split out to a separate patch with its own justification

File: sk.c [hunk 5]
```diff
@@ -360,22 +368,22 @@ int sk_receive(int fd, void *buf, int buflen,
...
-		memset(&hwts->ts, 0, sizeof(hwts->ts));
-		return cnt;
-	}
-
 	switch (hwts->type) {
 	case TS_SOFTWARE:
-		hwts->ts = timespec_to_tmv(ts[0]);
+		if (ts)
```
Comment (mcb30) [inline sk.c:374]: Conditional will not be needed if the "if(!ts)" block remains in place

File: sk.c [hunk 6]
```diff
@@ -360,22 +368,22 @@ int sk_receive(int fd, void *buf, int buflen,
...
 	case TS_P2P1STEP:
-		hwts->ts = timespec_to_tmv(ts[2]);
+		if (hr)
+			hwts->ts = timehires_to_tmv(hr[0]);
+		else if (ts)
+			hwts->ts = timespec_to_tmv(ts[2]);
 		break;
 	case TS_LEGACY_HW:
-		hwts->ts = timespec_to_tmv(ts[1]);
+		if (ts)
```
Comment (mcb30) [inline sk.c:386]: Conditional will not be needed if the "if(!ts)" block remains in place

File: sk.c [hunk 7]
```diff
@@ -360,22 +368,22 @@ int sk_receive(int fd, void *buf, int buflen,
...
+			hwts->ts = timespec_to_tmv(ts[0]);
 		break;
 	case TS_HARDWARE:
 	case TS_ONESTEP:
 	case TS_P2P1STEP:
-		hwts->ts = timespec_to_tmv(ts[2]);
+		if (hr)
+			hwts->ts = timehires_to_tmv(hr[0]);
+		else if (ts)
```
Comment (mcb30) [inline sk.c]: Can just be "else" if the "if(!ts)" block remains in place

File: tmv.c [hunk 8]
```diff
@@ -25,7 +25,8 @@
 #include "msg.h"
 #include "tmv.h"
 
-#define NS_FRAC 0x10000
```
Comment (mcb30) [inline tmv.c]: These three lines may as well become part of the (not yet upstreamed) "tmv: Include fractional nanoseconds within internal representation".  I will merge them into that commit and force-push the "subns" branch

**Comment by pmattila** [comment]
Makes lots of sense. Fixes in pma/ns_next.
